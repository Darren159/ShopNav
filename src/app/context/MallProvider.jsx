import { createContext, useState, useEffect, useMemo } from "react";
import { Alert } from "react-native";
import PropTypes from "prop-types";
import fetchMalls from "../../services/fetchMalls";

export const MallContext = createContext();

export function MallProvider({ children }) {
  const [malls, setMalls] = useState([]);
  const [currentMall, setCurrentMall] = useState(null);

  useEffect(() => {
    const initializeMalls = async () => {
      try {
        const mallList = await fetchMalls();
        setMalls(mallList);
        setCurrentMall(mallList[0]); // Set the first mall as the current mall
      } catch (error) {
        Alert.alert("Error", error.message, [{ text: "OK" }], {
          cancelable: false,
        });
      }
    };
    initializeMalls();
  }, []);
  const value = useMemo(
    () => ({ malls, currentMall, setCurrentMall }),
    [malls, currentMall]
  );

  return <MallContext.Provider value={value}>{children}</MallContext.Provider>;
}

MallProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
