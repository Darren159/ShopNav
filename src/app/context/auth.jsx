import { createContext, useState, useEffect, useMemo } from "react";
import { Alert } from "react-native";
import {
  onAuthStateChanged,
  signInWithEmailAndPassword,
  signOut,
} from "firebase/auth";
import PropTypes from "prop-types";

import { auth } from "../../../firebaseConfig";

export const AuthContext = createContext({});

export function AuthProvider({ children }) {
  const [user, setUser] = useState(null);

  useEffect(() => {
    onAuthStateChanged(auth, (userState) => {
      if (userState) {
        setUser(userState);
      } else {
        setUser(null);
      }
    });
  }, []);

  const authContextValue = useMemo(
    () => ({
      user,
      setUser,
      signin: async (email, password) => {
        try {
          await signInWithEmailAndPassword(auth, email, password);
        } catch (error) {
          let errorMessage;
          
          switch (error.code) {
            case 'auth/user-not-found':
              errorMessage = 'There is no user record corresponding to this identifier. The user may have been deleted.';
              break;
            case 'auth/wrong-password':
              errorMessage = 'The password is invalid or the user does not have a password.';
              break;
            case 'auth/invalid-email':
              errorMessage = 'The email address is badly formatted.';
              break;
            default:
              errorMessage = 'Something went wrong. Please try again later.';
          }
      
          Alert.alert(
            "Authentication Failed",
            errorMessage,
            [
              { text: "OK" }
            ]
          );
        }
      },
      signout: async () => {
        try {
          await signOut(auth);
        } catch (e) {
          console.error(e);
        }
      },
    }),
    [user]
  );

  return (
    <AuthContext.Provider value={authContextValue}>
      {children}
    </AuthContext.Provider>
  );
}

AuthProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
