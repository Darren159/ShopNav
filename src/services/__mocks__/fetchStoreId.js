const mockFetchStoreId = jest.fn();

export default mockFetchStoreId;
