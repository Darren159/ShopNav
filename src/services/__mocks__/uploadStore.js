const mockUploadStore = jest.fn();

export default mockUploadStore;
