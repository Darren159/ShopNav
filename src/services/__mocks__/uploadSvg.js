const mockUploadSvg = jest.fn();

export default mockUploadSvg;
