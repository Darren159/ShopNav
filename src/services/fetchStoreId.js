import { doc, getDoc } from "firebase/firestore";
import { db } from "../../firebaseConfig";

// This asynchronous function fetches the ID of a store in a given mall based on its name. 
// Parameters: 
// currentMall - The ID or name of the mall where the store is located
// storeName - The name of the store whose ID needs to be fetched
export default async function fetchStoreId(currentMall, storeName) {

  // Format the store name by replacing spaces with hyphens and converting to lowercase
  const formattedStoreName = `${storeName.replace(/\s/g, "-").toLowerCase()}`;

  // Construct two potential document IDs (the combination of mall and store names can be in two orders)
  const documentID1 = `${currentMall.toLowerCase()}-${formattedStoreName}`;
  const documentID2 = `${formattedStoreName}-${currentMall.toLowerCase()}`;

  // Create a reference to the first potential document
  const docRef1 = doc(db, "malls", currentMall, "stores", documentID1);
  const docSnap1 = await getDoc(docRef1);

  // If the first document exists, return its ID
  if (docSnap1.exists()) {
    return docSnap1.id;
  }

  // Create a reference to the second potential document
  const docRef2 = doc(db, "malls", currentMall, "stores", documentID2);
  const docSnap2 = await getDoc(docRef2);

  // If the second document exists, return its ID
  if (docSnap2.exists()) {
    return docSnap2.id;
  }

   // If neither document exists (i.e., the store name is invalid), throw an error
  throw new Error("Invalid Store Name");
}
