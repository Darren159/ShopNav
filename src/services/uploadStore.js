import { httpsCallable } from "firebase/functions";
import { functions } from "../../firebaseConfig";

export default async function uploadStore(currentMall, storeDocId, promoInfo) {
  const uploadStoreData = httpsCallable(functions, "uploadStoreData");
  const storeResponse = await uploadStoreData({
    mall: currentMall,
    store: storeDocId,
    promo: promoInfo,
  });
  if (!storeResponse.data.success) {
    throw new Error(
      "Failed to Upload Store Data. Please check your internet connection"
    );
  }
}
