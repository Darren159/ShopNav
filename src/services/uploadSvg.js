import { httpsCallable } from "firebase/functions";
import * as FileSystem from "expo-file-system";
import { functions } from "../../firebaseConfig";

// Function to upload SVG data for a specific mall
// Parameters: currentMall - the mall for which the SVG data is being uploaded
//             svgUri - the URI where the SVG data is stored
//             filename - name of the SVG file being uploaded
export default async function uploadSvg(currentMall, svgUri, filename) {
  let svgString;
  try {

     // Attempt to read the SVG file as a string
    svgString = await FileSystem.readAsStringAsync(svgUri);
  } catch (error) {
    throw new Error("Failed to read file. Please try again.");
  }

  // Define the Cloud Function for uploading SVG data
  const uploadSVGData = httpsCallable(functions, "uploadSVGData");

  // Call the Cloud Function with the SVG string and the mall name
  const svgResponse = await uploadSVGData({
    svg: svgString,
    mall: currentMall,
  });

  // Check the response from the Cloud Function. If unsuccessful, throw an error
  if (!svgResponse.data.success) {
    throw new Error(
      "Failed to Upload SVG Data. Please check that you have the correct file."
    );
  }

  // Define the Cloud Function for uploading mall layout
  const uploadMallLayout = httpsCallable(functions, "uploadMallLayout");

  // Call the Cloud Function with the SVG string, the mall name, and the filename
  const layoutResponse = await uploadMallLayout({
    svg: svgString,
    mall: currentMall,
    filename,
  });

  // Check the response from the Cloud Function. If unsuccessful, throw an error
  if (!layoutResponse.data.success) {
    throw new Error(
      "Failed to Upload Mall Layout. Please check that you have the correct file"
    );
  }
}
