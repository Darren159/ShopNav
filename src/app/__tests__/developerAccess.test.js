import { render, fireEvent, waitFor } from "@testing-library/react-native";
import renderer from "react-test-renderer";
import { Alert } from "react-native";
import * as DocumentPicker from "expo-document-picker";
import { MallContext } from "../context/mallProvider";
import DeveloperAccess from "../developerAccess";
import fetchStoreId from "../../services/fetchStoreId";
import uploadStore from "../../services/uploadStore";
import uploadSvg from "../../services/uploadSvg";

jest.mock("expo-document-picker");

jest.mock("../../services/fetchStoreId");

jest.mock("../../services/fetchMalls");

jest.mock("../../services/uploadSvg");

jest.mock("../../services/uploadStore");

const setCurrentMallMock = jest.fn();

const mockedContextValue = {
  currentMall: "Mall 1",
  setCurrentMall: setCurrentMallMock,
  malls: ["Mall 1", "Mall 2", "Mall 3"],
};

jest.spyOn(Alert, "alert");

DocumentPicker.getDocumentAsync.mockResolvedValue({
  type: "success",
  assets: [
    {
      name: "test-file.svg",
      uri: "file:/test-file.svg",
    },
  ],
});

describe("DeveloperAccess", () => {
  it("renders correctly", async () => {
    const tree = renderer
      .create(
        <MallContext.Provider value={mockedContextValue}>
          <DeveloperAccess />
        </MallContext.Provider>
      )
      .toJSON();
    await waitFor(() => expect(tree).toMatchSnapshot());
  });

  it("handles file selection", async () => {
    const { getByText } = render(
      <MallContext.Provider value={mockedContextValue}>
        <DeveloperAccess />
      </MallContext.Provider>
    );

    fireEvent.press(getByText("Select SVG"));
    expect(DocumentPicker.getDocumentAsync).toHaveBeenCalled();

    await waitFor(() =>
      expect(getByText("Selected file: test-file.svg")).toBeTruthy()
    );
  });

  it("handles file upload", async () => {
    uploadSvg.mockResolvedValue();

    const { getByText } = render(
      <MallContext.Provider value={mockedContextValue}>
        <DeveloperAccess />
      </MallContext.Provider>
    );

    fireEvent.press(getByText("Select SVG"));
    await waitFor(() =>
      expect(DocumentPicker.getDocumentAsync).toHaveBeenCalled()
    );

    fireEvent.press(getByText("Upload SVG"));

    await waitFor(() =>
      expect(uploadSvg).toHaveBeenCalledWith(
        "Mall 1",
        "file:/test-file.svg",
        "test-file.svg"
      )
    );
  });

  it("handles store information upload", async () => {
    fetchStoreId.mockResolvedValue("store-id");
    uploadStore.mockResolvedValue();

    const { getByText, getByPlaceholderText } = render(
      <MallContext.Provider value={mockedContextValue}>
        <DeveloperAccess />
      </MallContext.Provider>
    );

    fireEvent.changeText(getByPlaceholderText("Enter Store"), "Test Store");
    fireEvent.changeText(
      getByPlaceholderText("Input Promo Info"),
      "Promo Info"
    );
    fireEvent.press(getByText("Upload Store Info"));

    expect(fetchStoreId).toHaveBeenCalledWith("Mall 1", "Test Store");

    await waitFor(() =>
      expect(uploadStore).toHaveBeenCalledWith(
        "Mall 1",
        "store-id",
        "Promo Info"
      )
    );
  });
});
