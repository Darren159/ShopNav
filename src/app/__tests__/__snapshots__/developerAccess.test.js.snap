// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeveloperAccess renders correctly 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "additive",
      "left": "additive",
      "right": "additive",
      "top": "additive",
    }
  }
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "borderColor": "grey",
        "borderRadius": 5,
        "borderWidth": 1,
        "margin": 10,
        "padding": 10,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 18,
          "fontWeight": "bold",
        }
      }
    >
      SVG
    </Text>
    <View
      style={
        {
          "padding": 10,
        }
      }
    >
      <Text>
         
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#8AB9EF",
              "borderColor": "grey",
              "borderRadius": 10,
              "borderWidth": 1,
              "marginTop": 5,
              "padding": 10,
              "width": 300,
            }
          }
        >
          <Text
            style={
              {
                "color": "white",
              }
            }
          >
             
            Select SVG
             
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        {
          "alignItems": "center",
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#8AB9EF",
              "borderColor": "grey",
              "borderRadius": 10,
              "borderWidth": 1,
              "marginTop": 5,
              "padding": 10,
              "width": 300,
            }
          }
        >
          <Text
            style={
              {
                "color": "white",
              }
            }
          >
             
            Upload SVG
             
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "borderColor": "grey",
        "borderRadius": 5,
        "borderWidth": 1,
        "margin": 10,
        "padding": 10,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 18,
          "fontWeight": "bold",
        }
      }
    >
      Store Info
    </Text>
    <View
      style={
        {
          "padding": 10,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#fff",
            "borderRadius": 10,
            "borderWidth": 1,
            "flexDirection": "row",
            "paddingLeft": 5,
          }
        }
      >
        <Text />
        <TextInput
          onChangeText={[Function]}
          placeholder="Enter Store"
          style={
            {
              "flex": 1,
              "paddingLeft": 5,
            }
          }
          value=""
        />
      </View>
    </View>
    <TextInput
      onChangeText={[Function]}
      placeholder="Input Promo Info"
      style={
        {
          "backgroundColor": "#fff",
          "borderRadius": 10,
          "borderWidth": 1,
          "marginHorizontal": 10,
          "paddingLeft": 10,
          "paddingVertical": 3,
          "textAlign": "left",
        }
      }
      value=""
    />
    <View
      style={
        {
          "alignItems": "center",
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#8AB9EF",
              "borderColor": "grey",
              "borderRadius": 10,
              "borderWidth": 1,
              "marginTop": 5,
              "padding": 10,
              "width": 300,
            }
          }
        >
          <Text
            style={
              {
                "color": "white",
              }
            }
          >
             
            Upload Store Info
             
          </Text>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaView>
`;
